File: algorithms/508. Most Frequent Subtree Sum/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findFrequentTreeSum(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        def dfs(node):
            if not node:
                return 0
            s = node.val + dfs(node.left) + dfs(node.right)
            counter[s] += 1
            return s

        counter = collections.defaultdict(int)
        dfs(root)
        max_cnt = max(counter.itervalues()) if counter else 0
        return [k for k, v in counter.iteritems() if v == max_cnt]

File: algorithms/508. Most Frequent Subtree Sum/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> findFrequentTreeSum(TreeNode* root) {
        map<int, int> count;
        dfs(root, count);
        int m = 0;
        for ( auto & p : count ) {
            m = max(m, p.second);
        }
        vector<int> result;
        for ( auto & p : count ) {
            if ( p.second == m ) {
                result.push_back(p.first);
            }
        }
        return result;
    }

    int dfs(TreeNode * root, map<int, int> & count) {
        if ( ! root ) {
            return 0;
        }
        int sum = dfs(root->left, count) + dfs(root->right, count) + root->val;
        ++count[sum];
        return sum;
    }
};
